### ISSUE TITLE
iCade 8Bitdo FC30 Controller : Bad mapping

### ISSUE THREAD

**Issue by Canadadry** [jasarien/Provenance#601]
Hi I've tried to play with a 8Bitdo FC30 Controller on my iPhone IOS10 (French). The dpad is not mapped correctly I cant move in any direction but other button are working fine. 

So I dig up a little and I found my controller is not sending the same key as defined in iCadeReaderView.h

```
/*
  What I should have => what I really have
 UP ON,OFF  = w,e    => z,e
 RT ON,OFF  = d,c    => d,c
 DN ON,OFF  = x,z    => x,w
 LT ON,OFF  = a,q    => q,a
 A  ON,OFF  = y,t    => y,t (Y)
 B  ON,OFF  = h,r    => h,r (B)
 C  ON,OFF  = u,f    => u,f (A)
 D  ON,OFF  = j,n    => j,n (X)
 E  ON,OFF  = i,m    => i,[,] (R)
 F  ON,OFF  = k,p    => k,p (L)
 G  ON,OFF  = o,g    => o,g (Select)
 H  ON,OFF  = l,v    => l,v (Start)
*/
```

It seems that the controller is sending keychar as a AZERTY keyboard (since my iphone is in french) and provenance expect the keychar form an QWERTY keyboard. 

I guess we should extract the mapping as localisable data. 

For now I have edited iCadeRedearView.m and it work fine, but that cannot be a patch.  Here what I have changed to make it work. 

```
//static const char *ON_STATES  = "wdxayhujikol"; //wdxazhujikol for German keyboard layout
//static const char *OFF_STATES = "eczqtrfnmpgv"; //ecyqtrfnmpgv for German keyboard layout

static const char *ON_STATES  = "zdxqyhujikol"; //wdxazhujikol for French keyboard layout
static const char *OFF_STATES = "ecwatrfn,pgv"; //ecyqtrfnmpgv for French keyboard layout
```

**Comment by Vahron** [comment]
Hi,

Just to confirm I have the exact same issue with a SFC30 controller (it's the same as the FC30 and has the same mapping, only its form factor is different) on a French iPhone 7 (iOS 11). Your workaround will help while waiting for a proper fix, thank you!

**Comment by CodyJLeeApps** [comment]
Thinking about this outside the context of this app:
Based on my understanding of the 8Bitdo Controller (and the company instructions) the keyboard you have activated in the device needs to be “English (US)” for the controller to work properly.

I have used this app extensively with the 8Bitdo Controller and have had no issues.

Try setting your virtual keyboard to English (US) and give it another try.

**Comment by CodyJLeeApps** [comment]
Thinking about this inside the context of the app, though: 
I haven't spent much time in the code, so I will have to investigate... It would be nice if we had a struct inside the applicable class which essentially allowed the following:

1. Check current device type
2. Check localization 
3. use appropriate localization map (constants) for the iCade Reader.

**Comment by Canadadry** [comment]
I am not sure we have that kind of setting in iOS. 

Have you found where it can be changed ?

**Comment by CodyJLeeApps** [comment]
@Canadadry I will add some instructions for the temporary re-assignment of your keyboard, just a couple minutes.

**Comment by CodyJLeeApps** [comment]
1. Settings to open keyboard options
![1_settings](https://user-images.githubusercontent.com/18606509/31494672-65c6f096-af22-11e7-85f3-b32f52747a69.jpg)

2. Current keyboards
![2_keyboards_before](https://user-images.githubusercontent.com/18606509/31494689-748a7bca-af22-11e7-8bdc-0a9eebfb1b11.jpg)

3. After adding keyboard
![3_keyboards_after](https://user-images.githubusercontent.com/18606509/31494695-785a68be-af22-11e7-82f3-4d8140c46219.jpg)


4. selecting keyboard, use any app that lets you see the keyboard (you can't have the 8bitdo controller connected yet)press and hold the "world" button and the keyboard options pop up.
![4_keyboards_changing](https://user-images.githubusercontent.com/18606509/31494699-7b7db32a-af22-11e7-8edb-314fca4f0701.jpg)

**Comment by CodyJLeeApps** [comment]
Let me know if anything is unclear

**Comment by Vahron** [comment]
@CodyJLeeApps It does work, but only if you make sure the English keyboard is set as the first in the list.

It's a valid workaround though, so many thanks for this!

**Comment by CodyJLeeApps** [comment]
@Vahron I didn't realize it had an ordering concern, thanks for mentioning that!

**Comment by Canadadry** [comment]
@CodyJLeeApps don't you think it's a bit intrusive ? 

And what if I must have a French keyboard for another app ?

**Comment by CodyJLeeApps** [comment]
@Canadadry I do think its intrusive, although it's a valid workaround until the functionality is added to the application.

**Comment by Namavila** [comment]
Please help! I just bought SFC30 8bitdo and I discovered that poorly works with almost IOS SNES emulators. I didnt know about iCade and MFI, whatever. Searching, Provenance came up and I see you use those 8bitdo controller! YES! :) But I could not make it work yet, I dont know what is missing. Can anyone help? It just dont work, like it wasnt connected. I chosse 8bitdo on "iCade controller", and then try do make any press be recognized in "Controller Selection", but nothing happens and keeps "None Selected". In the game happens the same. I garantee the bluetooth is conected, and the light is solid blue on the game pad. What am I missing? Thank you in advance.

**Comment by larssteenhoff** [comment]
On iOS 11 for the Apple TV the mappings are also messed up, don't know how to fix it

**Comment by jasarien** [comment]
Should be fixed with PR #614

**Comment by Canadadry** [comment]
I am sorry to disappoint you but no this is not working. 
There is a macro enabling the patch only on OS_TV, but for other plateform they stick to english. 

Here one of the modification of the plugin. 

``` 
(void)applyKeyMappingForCurrentLocale {
 #if TARGET_OS_TV
     NSString *localeIdentifier = [[NSLocale currentLocale] localeIdentifier];
     if ([localeIdentifier hasPrefix:@"de"]) {
         self.onStates = ON_STATES_DE;
         self.offStates = OFF_STATES_DE;
     } else if ([localeIdentifier hasPrefix:@"fr"]) {
         self.onStates = ON_STATES_FR;
         self.offStates = OFF_STATES_FR;
     } else {
         self.onStates = ON_STATES_EN;
         self.offStates = OFF_STATES_EN;
     }
 #else
     self.onStates = ON_STATES_EN;
     self.offStates = OFF_STATES_EN;
 #endif
 }
```

I dont see why there a `#if`
